### PR TITLE
Update CocospoonLocalizerTest to expose the cocospoon problem

### PR DESCRIPTION
**Pull request by DehengYang**
This test aims to expose the java.lang.RuntimeException when running test cases in Math_58, a real bug from defects4j. But this test file (CocospoonLocalizerTest.java) is created based on the version of Nopol of March 2017.

### CODE REVIEW

**Comment by monperrus** [comment]
We make good progress. Instead of modifying an existing test, could only add a new test?  Thanks.

File: nopol/src/test/java/fr/inria/lille/localization/CocospoonLocalizerTest.java [hunk 1]
```diff
@@ -11,93 +11,38 @@
...
  */
 public class CocospoonLocalizerTest {
 
     @Test
     public void testOchiaiCoCoSpoonLocalizer() throws Exception {
-
-		/* test OchiaiCoCoSpoonLocalizer : the SourceLocation must be sorted following the Ochiai metric */
-
-        File[] sources = new File[]{new File("../test-projects/src/main/java/nopol_examples/nopol_example_1/NopolExample.java")};
+        File[] sources = new File[]{new File("/home/apr/environment/buggyPrograms/NopolRepairPrograms/Math/Math_58/src/main/java")}; // This is the source file of Math_58
```
Comment (monperrus) [inline nopol/src/test/java/fr/inria/lille/localization/CocospoonLocalizerTest.java]: no absolute path in tests

**Comment by DehengYang** [comment]
I have created a new test named `CocospoonLocalizer2Test`. Thank you!

**Comment by monperrus** [comment]
Thanks, we make progress.

Could you make sure that the new test compiles and fails? (see Travis log
https://travis-ci.org/SpoonLabs/nopol/jobs/487725670)

**Comment by DehengYang** [comment]
I am sorry that I reply so late these days. Since today is Chinese New Year's day, which is a big festival in China, I have spent much time in family reunion these days.

Thank you for your understanding!

After I updated this test just now, the test can actually compile.  Now the build log is (see https://travis-ci.org/SpoonLabs/nopol/jobs/488800318#L4389):

image

The reason is that the defects4j bug Math 58 does not exist, which can actually be downloaded using `defects4j checkout` command. Should I upload this Math 58 bug dir now?

**Comment by DehengYang** [comment]
After I uploaded Math_58 dir, I found that it seems that Math 58 cannot compile in jdk 1.8. The info is as follows:
image

**Comment by DehengYang** [comment]
After the new commits above, the running info is (see https://travis-ci.org/SpoonLabs/nopol/jobs/488938376):

> Running fr.inria.lille.localization.CocospoonLocalizer2Test
> Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 38.689 sec <<< FAILURE!
> testOchiaiCoCoSpoonLocalizer(fr.inria.lille.localization.CocospoonLocalizer2Test)  Time elapsed: 38.689 sec  <<< FAILURE!
> java.lang.AssertionError: 
> 	at gov.nasa.jpf.util.test.TestJPF.fail(TestJPF.java:167)
> 	at gov.nasa.jpf.util.test.TestJPF.assertTrue(TestJPF.java:1176)
> 	at gov.nasa.jpf.util.test.TestJPF.assertTrue(TestJPF.java:1181)
> 	at fr.inria.lille.localization.CocospoonLocalizer2Test.testOchiaiCoCoSpoonLocalizer(CocospoonLocalizer2Test.java:41)

Actually this is out of my expectation, because the `java.lang.RuntimeException` (see https://github.com/SpoonLabs/nopol/issues/179) does not occur in this process.

I will keep exploring the reason behind this result. Thank you so much for your help and continuous guidance!

**Comment by monperrus** [comment]
Instead of pushing all Math_58/ to the repo, you should ask travis to download it, as done in https://github.com/SpoonLabs/nopol/blob/master/nopol/src/test/java/fr/inria/lille/repair/nopol/Defects4jUtils.java#L41

Then, your pull-request contains one single changed file, not 2079 ones :-)

**Comment by DehengYang** [comment]
Thank you so much for your great help to me! And I am really sorry for my lack of experience in this aspect.

Besides this, I found that in travis `CocospoonLocalizer2Test` that I wrote cannot trigger the failure (i.e., `java.lang.RuntimeException` mentioned before) which previously occured in my lab computer.

I think more efforts should be put on triggering the expected failure. However, several days ago, my lab computers were shut down because of the vocation. I will operate on my lab computer later to explore the reasons behind.

Thank you again for your great kindness!

**Comment by DehengYang** [comment]
I originally planned to operate on my lab PC, but I cannot wait to explore the reasons behind. Therefore, I reconfigure CoCoSpoon, Nopol, Defects4J, and jdk versions (1.7 and 1.8) in my Ubuntu cloud server.

After I have made a series of experiments, I think I find the reasons. 
1) When the CoCoSpoon is installed in jdk 1.7 and the Nopol is packaged in jdk 1.7, running `CocospoonLocalizer2Test` with command ` mvn test -Dtest=CocospoonLocalizer2Test` will trigger the expected `java.lang.RuntimeException` (mentioned in https://github.com/SpoonLabs/nopol/issues/179 ).
2) When the CoCoSpoon is installed in jdk 1.8 and the Nopol is packaged in jdk 1.8, running `CocospoonLocalizer2Test` won't trigger the expected  `java.lang.RuntimeException` .

Therefore, the bug I found in https://github.com/SpoonLabs/nopol/issues/179 actually only exists in old version of CoCoSpoon (under jdk 1.7).

Personally, though this is the bug of the old version, such bug still deserves to be solved, as current defects4j benchmark mainly supports jdk 1.7.  For example, the mostly used java benchmark in automated program repair is defects4j version 1.1.0, which explicitly only supports jdk 1.7.

This is my conclusion. Thank you for your help these days, which really stimulates me to make further attempts! So... should I cancel this pull request now, or make further modifications? 

Thank you again for your continuous guidance! That really helps me a lot !

**Comment by monperrus** [comment]
Thanks for the progress.

The goal is that Travis CI reproduces the failure you're observing.

Right now, the failure in Travis is:

    spoon.compiler.InvalidClassPathException: /home/travis/build/SpoonLabs/nopol/nopol/../Math_58/target/classes does not exist

See https://travis-ci.org/SpoonLabs/nopol/jobs/492487049

It seems your checkout of Math_58 is not correct.

Could you update it?

**Comment by DehengYang** [comment]
I have updated it now, and this running info of `CocospoonLocalizer2Test` in `travis_core.sh` is:
>Running fr.inria.lille.localization.CocospoonLocalizer2Test
>...
>...
>...
>Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 71.665 sec

This test passes in travis without triggering `RuntimeException` because the test is running in JDK 1.8. As I pointed out in the `CocospoonLocalizer2Test.java`:

> /**
> \* Modified by deheng on Feb 1, 2019.
> \* This test aims to expose the java.lang.RuntimeException when running test cases in Math_58, a real bug from defects4j. But this test file (CocospoonLocalizerTest.java) is created based on the version of Nopol of March 2017.
> \* Use `mvn clean package -DskipTests` to build nopol and use `mvn test -Dtest=CocospoonLocalizerTest` to run this test.
> \*/
>
> // But this java.lang.RuntimeException seems to only happen when the Nopol and CoCospoon are built/installed in JDK 1.7. When Nopol and CoCoSpoon are built/installed in JDK 1.8, this test will pass.

That's all I have explored about this issue.

Thank you for your fast reply and great help!

**Comment by monperrus** [comment]
We need to reproduce the bug to understand and fix it.

You can update `.travis_core.sh` to force using Java7 (by using `jdk_switcher`, see Travis doc).

Thanks.

**Comment by DehengYang** [comment]
Thank you for your help! Now I have switched the jdk version to 1.7 by modifying `.travis_core.sh` and `.travis.yml`. 

However, the travis build process failed. The failure info is as follow:

>[ERROR] Failed to execute goal on project nopol: Could not resolve dependencies for project fr.inria.gforge.spirals:nopol:jar:0.2-SNAPSHOT: Failed to collect dependencies at fil.iagl.cocospoon:CocoSpoon:jar:1.0.0-SNAPSHOT: Failed to read artifact descriptor for fil.iagl.cocospoon:CocoSpoon:jar:1.0.0-SNAPSHOT: Could not transfer artifact fil.iagl.cocospoon:CocoSpoon:pom:1.0.0-SNAPSHOT from/to tdurieux.github.io/maven-repository/snapshots/ (https://tdurieux.github.io/maven-repository/snapshots/): java.security.ProviderException: java.security.InvalidKeyException: EC parameters error: no such provider: SunEC -> [Help 1]

I then made further attempts to understand the failure, such as removing the repo ` tdurieux.github.io/maven-repository/snapshots/` and run the Nopol in my Ubuntu Cloud Server (where the cocospoon is installed in the local Maven repo). 

Finally, I found that the current Nopol version is quite different from the buggy version of Nopol that I used in March 2017. For example, when I configure the current Nopol version in jdk 1.7 and run `mvn test -Dtest=CocospoonLocalizer2Test`, the error happens as follow:

> -------------------------------------------------------
>  T E S T S
> -------------------------------------------------------
> Running fr.inria.lille.localization.CocospoonLocalizer2Test
> Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.792 sec <<< FAILURE!
> testOchiaiCoCoSpoonLocalizer(fr.inria.lille.localization.CocospoonLocalizer2Test)  Time elapsed: 0.231 sec  <<< ERROR!
> java.lang.UnsupportedClassVersionError: spoon/processing/Processor : Unsupported major.minor version 52.0
> 	at java.lang.ClassLoader.defineClass1(Native Method)
> 	at java.lang.ClassLoader.defineClass(ClassLoader.java:800)
> 	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
> 	at java.net.URLClassLoader.defineClass(URLClassLoader.java:449)
> 	at java.net.URLClassLoader.access$100(URLClassLoader.java:71)
> 	at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
> 	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
> 	at java.security.AccessController.doPrivileged(Native Method)
> 	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
> 	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
> 	at fr.inria.lille.localization.CocospoonLocalizer2Test.testOchiaiCoCoSpoonLocalizer(CocospoonLocalizer2Test.java:58)

This means, the `spoon` library used in Nopol has changed a lot (see http://spoon.gforge.inria.fr/repositories/snapshots/fr/inria/gforge/spoon/spoon-core/ and http://spoon.gforge.inria.fr/).  For example, the old Nopol uses `fr.inria.gforge.spoon.spoon-core 5.5.0 version` while the current Nopol uses `fr.inria.gforge.spoon.spoon-core 7.0.0 version`, which makes the reproducing process harder than I thought. Thus, this bug I reported  before cannot be reproduced by adding only one file. 

In my view, the `RuntimeException` bug may be caused by `cocospoon` or `spoon-core`. Additionaly, I think this `RuntimeException` bug I reported has already disappeared since both `Nopol` and `spoon lib` are continuously developed. **_But for the previous users like me who still run Nopol and cocospoon, and defects4j in jdk 1.7, this is a real problem._** 

In summary, I failed to reproduce the bug in current Nopol. I think to reproduce this bug, we need:
1) The old version of Nopol (e.g., the Nopol version proposed in March 2017) ;
2) The old version of CoCoSpoon (e.g., the CoCospoon proposed in March 2017);
3) The jdk 1.7.

This is my report after some experiments.  Thank you so much for your time! Thank you again for your great patience!

**Comment by monperrus** [comment]
> But for the previous users like me who still run Nopol and cocospoon, and defects4j in jdk 1.7, this is a real problem.

Note that the goal of Nopol is to be a program repair system for modern code, it is not to repair Defects4j only.

To further understand your problem, could you please update your pull request against branch "march2017" (and not against "master")?

**Comment by monperrus** [comment]
Closing, see explanation in #182
